### PR TITLE
S100/S101/S3909: Add support for record structs

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs:136]: There is a `ClassDeclarationSyntax` check in line 166:
`if (typeDeclaration is ClassDeclarationSyntax && IsTestClassName(typeDeclaration.Identifier.ValueText))`

https://github.com/SonarSource/sonar-dotnet/blob/a10e7dc4555267a726ff334ff654481a472fb5ab/analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs#L166-L169

Are there any test frameworks with support for `record` and/or `record struct` as test method containers (It just works probably)? Is anybody doing this?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionsShouldImplementGenericInterface.CSharp10.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionsShouldImplementGenericInterface.CSharp10.cs:9]: You know that the message assertion and the `^^^^` markers are optional because they are already in the main test file, right?
It makes no harm having it, it's just to be clear that it is not required.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionsShouldImplementGenericInterface.cs:20]: I would actually name it something like `GenericFirst` and the previous `GenericLast` to explain a little what it tests. It would also remove the need to rename the numbers on the rest :)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:39]: It's a property, please make it single line.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:43]: I'd be also fine with having the `new()` on the previous line. Just to save space (it's boring scaffolding)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:65]: `typeDeclaration` is `c.Node`, that cannot be `null` => this change will improve coverage.
```suggestion
                    var implementedTypes = typeDeclaration.BaseList?.Types;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:95]: Small syntax-sugar simplification
```suggestion
                    issues?.ForEach(c.ReportIssue);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:88]: Line 85&86 can be merged to
```
if(SuggestGenericCollectionType(typeSymbol) is { } suggestedGenericType)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:89]: This is short, can be a single line

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:59]: This bail out can be merged with the next bailout. To have only one condition to do it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:80]: Lines 74-77: I don't like the `continue`. Let's change the `foreach` to this, because the `typeSyntax` is not used for anything else.
```
                    foreach (var typeSymbol in implementedTypes.Value.Select(x => c.SemanticModel.GetSymbolInfo(x.Type).Symbol?.GetSymbolType()).WhereNotNull())

```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:80]: When looking at it, we can actually significantly improve performance of this rule by doing this instead:
```
var containingType = (INamedTypeSymbol)c.ContainingSymbol;
foreach (var typeSymbol in containingType.Interfaces.Concat(new[] { containingType.BaseType }).WhereNotNull())
```
We can avoid re-querying the symbols from syntax, as we already know the outer symbol.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
We can delete quite some code and improve performance from the collection rule

**Comment by martin-strecker-sonarsource** [comment]
@pavel-mikula-sonarsource Read for review. Please also look at https://github.com/SonarSource/sonar-dotnet/pull/5571#pullrequestreview-945966763 as it still apllies.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:86]: Why not? We're doing this quite often when possible

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:58]: Why not?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:58]: We can do it like this, the variables are just renamings, nothing expensive is happening.
```
var typeDeclaration = (BaseTypeDeclarationSyntax)c.Node;
                    var implementedTypes = typeDeclaration.BaseList?.Types;
                    if (implementedTypes == null || c.IsRedundantPositionalRecordContext())
                    {
                        return;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:86]: Ah, indeed (the explanation was not visible before)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs:134]: I don't know. My guess is it would work just fine for `record`. And at the same time I don't see a reason why would somebody use that.

I would leave it as it is now. We can change it if we see an issue raised by users.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, please merge the `if` from the last opened conversation before merging

**Pull request merged**
